docs/openfaas-pro/iam/function-authentication.md

```diff
@@ -0,0 +1,275 @@
+            request = Net::HTTP::Post.new(uri.request_uri)
+            request.set_form_data(request_params)
+            request['Content-Type'] = 'application/x-www-form-urlencoded'
+
+            # Execute the request
+            response = http.request(request)
+
+            # Parse the response
+            result = JSON.parse(response.body)
+
+            # Check the response for errors
+            if response.code != '200'
+                puts "Error: #{response.code}"
+                raise "Failed to retrieve token: #{result['error_description'] || 'Unknown error'}"
+            end
+
+            # Return the access token
+            result['access_token']
+        end
+        ```
+
+3. Invoke an authenticated function
+
+    Invoke an authenticated function with the function access token.
+
+    === "curl"
+
+        ```bash
+        curl -i "https://gateway.example.com/function/env" \
+            -H "Authorization: Bearer $(cat ./function-token.txt)"
+        ```
+
+    === "ruby"
+
+        ```ruby
+        def invoke_function(token)
+            function_url = "https://gateway.example.com/function/env"
+
+            # Create URI object
+            uri = URI(function_url)
+
+            # Create the HTTP object
+            http = Net::HTTP.new(uri.host, uri.port)
+            http.use_ssl = true if uri.scheme == 'https'
+
+            # Create the GET request
+            request = Net::HTTP::Get.new(uri.request_uri)
+            request['Authorization'] = "Bearer: #{token}"
+
+            # Execute the request
+            response = http.request(request)
+
+            # Check the response for errors
+            if response.code != '200'
+                puts "Error: #{response.code}"
+                raise "Failed to invoke function: code #{response.code}, body: #{response.body}"
+            end
+
+            # Return response body
+            response.body
+        end
+        ```
+
+### Invoke authenticated functions with the faas-cli
+ 
+The `faas-cli invoke` command can be used to invoke authenticated functions:
+
+```bash
+echo "OpenFaaS" | faas-cli invoke figlet
+```
+
+You will only be able to invoke authenticated function if the CLI is authenticated with the gateway, see: [SSO with CLI](/openfaas-pro/sso/cli/)
+
+The CLI detects if a function needs authentication by calling the function a first time without any authorization header. If it receives a 401 response code the response is inspected to check if authentication with a function access token is required. If this is the case the CLI automatically retries the request with an access token.
+
+If you know a function needs authentication you can provide the optional `--auth` flag to skip the detection step. The CLI will immediately include the access token in the first call.
+
+## See also
+
+- [Introducing built-in authentication for OpenFaaS Functions](https://www.openfaas.com/blog/built-in-function-authentication/)
```

Comment:
I think this would be better at the top as a link to a tutorial

Reply:
The link has been moved to the top.